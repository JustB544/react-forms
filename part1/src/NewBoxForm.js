import React, {useState} from "react";

function NewBoxForm({createBox}) {
    const INITIAL_STATE = {backgroundColor: "#000000", width: 100, height: 100};
    const [formData, setFormData] = useState(INITIAL_STATE);
    function handleSubmit(evt){
        evt.preventDefault();
        createBox(formData);
        setFormData(INITIAL_STATE);
    }

    function handleChange(evt){
        const {name, value} = evt.target;
        setFormData(fData => ({...fData, [name]: value}));
    }
    return (
        <form style={{display: "flex", flexDirection: "column", width: "fit-content"}} onSubmit={handleSubmit}>
            <div>
                <label htmlFor="backgroundColor">Background color: </label>
                <input name="backgroundColor" type="color" value={formData.backgroundColor} onChange={handleChange}/>
            </div>
            <div>
                <label htmlFor="height">Height: </label>
                <input name="height" type="number" value={formData.height} onChange={handleChange}/>
            </div>
            <div>
                <label htmlFor="width">Width: </label>
                <input name="width" type="number" value={formData.width} onChange={handleChange}/>
            </div>
            <button>Create</button>
        </form>
    )
}

export default NewBoxForm;