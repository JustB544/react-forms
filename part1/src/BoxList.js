import React, {useState} from "react";
import Box from "./Box";
import {v4 as uuid} from 'uuid';
import NewBoxForm from "./NewBoxForm";

function BoxList(props) {
    const [boxes, setBoxes] = useState(props.initial || {});

    function deleteBox(id){
        setBoxes(({[id]: undefined, ...b}) => b);
    }

    function createBox({backgroundColor, height, width}){
        setBoxes(b => ({...b, [uuid()]: { backgroundColor, height, width }}));
    }

    return (
        <div>
            { Object.keys(boxes).map(b => <Box 
            key={b} 
            id={b}
            backgroundColor={boxes[b].backgroundColor} 
            height={boxes[b].height} 
            width={boxes[b].width} 
            deleteBox={evt => deleteBox(b)}/>) }
            <NewBoxForm createBox={(box, evt) => createBox(box)}/>
        </div>
    );
}

export default BoxList;