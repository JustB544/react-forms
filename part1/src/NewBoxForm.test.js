import { render, fireEvent } from "@testing-library/react";
import NewBoxForm from "./NewBoxForm";
import BoxList from "./BoxList";

it("renders", () => {
    render(<NewBoxForm />);
});
  
it("matches snapshot", () => {
    const {asFragment} = render(<NewBoxForm />);
    expect(asFragment()).toMatchSnapshot();
});

it("creates a box", () => {
    const {container} = render(<BoxList/>);
    const button = container.querySelector("form button");
    fireEvent.click(button);
    expect(container.querySelector(".Box")).toBeInTheDocument();
});