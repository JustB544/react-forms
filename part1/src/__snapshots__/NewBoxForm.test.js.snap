// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    style="display: flex; flex-direction: column;"
  >
    <div>
      <label
        for="backgroundColor"
      >
        Background color: 
      </label>
      <input
        name="backgroundColor"
        type="color"
        value="#000000"
      />
    </div>
    <div>
      <label
        for="height"
      >
        Height: 
      </label>
      <input
        name="height"
        type="number"
        value="100"
      />
    </div>
    <div>
      <label
        for="width"
      >
        Width: 
      </label>
      <input
        name="width"
        type="number"
        value="100"
      />
    </div>
    <button>
      Create
    </button>
  </form>
</DocumentFragment>
`;
